transports/quic/src/transport.rs

```diff
@@ -926,4 +942,86 @@ mod test {
             )
             .unwrap();
     }
+
+    /// - A listens on 0.0.0.0:0
+    /// - B listens on 127.0.0.1:0
+    /// - A dials B
+    /// - Source port of A at B is the A's listen port
+    #[allow(non_snake_case)]
+    #[cfg(feature = "tokio")]
+    #[tokio::test]
+    async fn test_local_listener_reuse() {
+        let keypair = libp2p_identity::Keypair::generate_ed25519();
+        let mut node_A = crate::tokio::Transport::new(Config::new(&keypair));
+
+        let keypair = libp2p_identity::Keypair::generate_ed25519();
+        let mut node_B = crate::tokio::Transport::new(Config::new(&keypair));
+
+        node_B
+            .listen_on(
+                ListenerId::next(),
+                "/ip4/127.0.0.1/udp/0/quic-v1".try_into().unwrap(),
+            )
+            .unwrap();
+
+        node_A
+            .listen_on(
+                ListenerId::next(),
+                "/ip4/0.0.0.0/udp/0/quic-v1".try_into().unwrap(),
+            )
+            .unwrap();
+
+        // node_A: wait until a listener reports a loopback address
+        poll_fn(|cx| {
+            let mut pinned = Pin::new(&mut node_A);
+            while let Poll::Ready(ev) = pinned.as_mut().poll(cx) {
+                if let TransportEvent::NewAddress { listen_addr, .. } = ev {
+                    if multiaddr_to_socketaddr(&listen_addr, false)
+                        .unwrap()
+                        .0
+                        .ip()
+                        .is_loopback()
+                    {
+                        return Poll::Ready(());
+                    }
+                }
+            }
+
+            Poll::Pending
+        })
+        .await;
+
+        let listener_addr = |node: &GenTransport<_>| {
+            node.listeners
+                .iter()
+                .next()
+                .unwrap()
+                .socket
+                .local_addr()
+                .unwrap()
+        };
+
+        let node_B_multiaddr =
+            socketaddr_to_multiaddr(&listener_addr(&node_B), ProtocolVersion::V1);
+
+        // node A dials node B
+        node_A.dial(node_B_multiaddr).unwrap().await.unwrap();
+
+        // Verify that node B received node A listening port
+        poll_fn(|cx| {
+            let mut pinned = Pin::new(&mut node_B);
+            while let Poll::Ready(e) = pinned.as_mut().poll(cx) {
```

Comment:
I wonder whether polling only `node_B` here is prone to race conditions. I.e. I would expect one having to poll both in case of any intermittent failure. (Though arguably, e.g. a packet loss is rather rare on localhost.)

Either way, I think you can simplify this test by using the `connect` method:

https://github.com/libp2p/rust-libp2p/blob/da94dcb23b0d65a0e94965e6e8ddb14899053bd8/transports/quic/tests/smoke.rs#L689-L709

What do you think of moving this test to `smoke.rs`?

Reply:
Moved the test to `smoke.rs` and simplified it, thanks! But due to listening on `0.0.0.0` it is a bit ugly.